linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker
+
+import scala.concurrent._
+
+import java.nio.charset.StandardCharsets.UTF_8
+
+import org.junit.Test
+import org.junit.Assert._
+
+import org.scalajs.ir.ClassKind
+import org.scalajs.ir.Names._
+import org.scalajs.ir.Trees._
+import org.scalajs.ir.Types._
+
+import org.scalajs.logging._
+
+import org.scalajs.junit.async._
+
+import org.scalajs.linker._
+import org.scalajs.linker.interface._
+
+import org.scalajs.linker.testutils._
+import org.scalajs.linker.testutils.TestIRBuilder._
+
+class IncrementalTest {
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  import IncrementalTest._
+
+  @Test // #4335
+  def testChangeMethodReachedFromExport(): AsyncResult = await {
+
+    val FooClass = ClassName("Foo")
+
+    val jsMethodName = StringLiteral("foo")
+    val staticMethodName = m("value", Nil, IntRef)
+
+    def classDefs(pre: Boolean) = Seq(
+        mainTestClassDef(
+            consoleLog(JSMethodApply(LoadModule(FooClass), jsMethodName, Nil))
+        ),
+        classDef(
+            FooClass,
+            kind = ClassKind.ModuleClass,
+            superClass = Some(ObjectClass),
+            memberDefs = List(
+                trivialCtor(FooClass),
+                JSMethodDef(EMF, jsMethodName, Nil,
+                    if (pre) int(5)
+                    else ApplyStatic(EAF, FooClass, staticMethodName, Nil)(IntType)
+                )(EOH, None),
+                MethodDef(EMF.withNamespace(MemberNamespace.PublicStatic),
+                    staticMethodName, NON, Nil, IntType, Some(int(6)))(EOH, None)
+            )
+        )
+    )
+
+    testIncremental(classDefs(_), _ => MainTestModuleInitializers)
+  }
+
+}
+
+object IncrementalTest {
+
+  def testIncremental(
+      classDefs: Boolean => Seq[ClassDef],
+      moduleInitializers: Boolean => List[ModuleInitializer],
+      config: StandardConfig = StandardConfig())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+    Future.sequence(Seq(
+        testIncrementalStep(backward = false, classDefs, moduleInitializers, config),
+        testIncrementalStep(backward = true, classDefs, moduleInitializers, config)
+    )).map(_ => ())
```

Comment:
Consider forcing the execution in order here, to reduce non-determinism in the test results:
```suggestion
    for {
        _ <- testIncrementalStep(backward = false, classDefs, moduleInitializers, config)
        _ <- testIncrementalStep(backward = true, classDefs, moduleInitializers, config)
    } yield ()
```

Reply:
I can do this if you want of course, but there isn't any non-determinism here (except for super fatal exceptions): `sequence` *does* return the results in order, it just executes them in parallel (which indeed might cause additional problems). Up to you.

Reply:
Ah so if both fail, it's always the failure of the first one that is reported? If yes, this is fine as  is.

Reply:
I have done it anyways because IMO it is much easier to understand.